### ISSUE TITLE
Notary errors when trying to add data persistence w/compose environment

### ISSUE THREAD

**Issue by mbentley** [docker/notary#322]
I've been doing some testing around using Content Trust (notary) with DTR 1.4 and one problem I am having is that the docker-compose.yml example does not allow for persistent data.  I tried to add persistence in https://github.com/mbentley/notary/commit/bfc6f6ef07d9028d919b9d18b8adde1273f19be4 but when I kill my compose stack and start it back up and then try to push another image to my DTR, I am seeing the notarysigner and notaryserver services throwing errors:

```
notaryserver_1 | time="2015-12-02T12:24:40Z" level=debug msg="200 GET" gun="dtr.mac/demo/myrepo" http.request.host=dtr.mac http.request.id=b2e1a249-3c66-4d28-b358-23e4c06dda86 http.request.method=GET http.request.remoteaddr=192.168.56.4 http.request.uri="/v2/dtr.mac/demo/myrepo/_trust/tuf/root.json" http.request.useragent="docker/1.9.1 go/go1.4.2 git-commit/a34a1d5 kernel/3.16.0-4-amd64 os/linux arch/amd64" tufRole=root
notaryserver_1 | time="2015-12-02T12:24:40Z" level=info msg="response completed" http.request.host=dtr.mac http.request.id=b2e1a249-3c66-4d28-b358-23e4c06dda86 http.request.method=GET http.request.remoteaddr=192.168.56.4 http.request.uri="/v2/dtr.mac/demo/myrepo/_trust/tuf/root.json" http.request.useragent="docker/1.9.1 go/go1.4.2 git-commit/a34a1d5 kernel/3.16.0-4-amd64 os/linux arch/amd64" http.response.duration=3.545051ms http.response.status=200 http.response.written=2377
notaryserver_1 | time="2015-12-02T12:24:40Z" level=debug msg="retrieving timestamp key for dtr.mac/demo/myrepo"
notaryserver_1 | time="2015-12-02T12:24:40Z" level=debug msg="sign called with 1 keys"
notarysigner_1 |
notarysigner_1 | (sql: Scan error on column index 1: unsupported driver -> Scan pair: []uint8 -> *time.Time)
notarysigner_1 | [2015-12-02 12:24:40]
notarysigner_1 |
notarysigner_1 | (sql: Scan error on column index 1: unsupported driver -> Scan pair: []uint8 -> *time.Time)
notarysigner_1 | [2015-12-02 12:24:40]
notarysigner_1 |
notarysigner_1 | (sql: Scan error on column index 1: unsupported driver -> Scan pair: []uint8 -> *time.Time)
notarysigner_1 | [2015-12-02 12:24:40]
notarysigner_1 |
notarysigner_1 | (sql: Scan error on column index 1: unsupported driver -> Scan pair: []uint8 -> *time.Time)
notarysigner_1 | [2015-12-02 12:24:40]
notarysigner_1 | time="2015-12-02T12:24:40Z" level=error msg="GetKeyInfo: key 77dde544cad17e911010eef98f4f15e137ab85bae5999754b9332281ae1a311a not found"
notaryserver_1 | time="2015-12-02T12:24:40Z" level=info msg="response completed" http.request.host=dtr.mac http.request.id=c367c202-56aa-44ae-a226-d3158e829f42 http.request.method=GET http.request.remoteaddr=192.168.56.4 http.request.uri="/v2/dtr.mac/demo/myrepo/_trust/tuf/timestamp.json" http.request.useragent="docker/1.9.1 go/go1.4.2 git-commit/a34a1d5 kernel/3.16.0-4-amd64 os/linux arch/amd64" http.response.duration=21.195973ms http.response.status=0 http.response.written=0
notaryserver_1 | 2015/12/02 12:24:40 http: panic serving 172.17.0.1:42949: runtime error: invalid memory address or nil pointer dereference
notaryserver_1 | goroutine 51 [running]:
notaryserver_1 | net/http.(*conn).serve.func1(0xc8205368f0, 0x7fb3f21925e8, 0xc820090000)
notaryserver_1 |        /usr/local/go/src/net/http/server.go:1287 +0xb5
notaryserver_1 | github.com/docker/notary/tuf/signed.Sign(0x7fb3f218c890, 0xc82023bb00, 0xc82049f680, 0xc820199228, 0x1, 0x1, 0x0, 0x0)
notaryserver_1 |        /go/src/github.com/docker/notary/tuf/signed/sign.go:54 +0x623
notaryserver_1 | github.com/docker/notary/server/timestamp.CreateTimestamp(0xc820566988, 0x13, 0xc82055d800, 0xc820402400, 0x1e8, 0x1e8, 0x7fb3f218ca08, 0xc82021da40, 0x7fb3f218c890, 0xc82023bb00, ...)
notaryserver_1 |        /go/src/github.com/docker/notary/server/timestamp/timestamp.go:142 +0x47f
notaryserver_1 | github.com/docker/notary/server/timestamp.GetOrCreateTimestamp(0xc820566988, 0x13, 0x7fb3f218ca08, 0xc82021da40, 0x7fb3f218c890, 0xc82023bb00, 0x0, 0x0, 0x0, 0x0, ...)
notaryserver_1 |        /go/src/github.com/docker/notary/server/timestamp/timestamp.go:75 +0x7cf
notaryserver_1 | github.com/docker/notary/server/handlers.GetTimestampHandler(0x7fb3f218c8e8, 0xc8204be2d0, 0x7fb3f2199b50, 0xc820566a40, 0xc8200d0a80, 0x0, 0x0)
notaryserver_1 |        /go/src/github.com/docker/notary/server/handlers/default.go:152 +0x5a9
notaryserver_1 | github.com/docker/notary/utils.(*rootHandler).ServeHTTP(0xc820279180, 0x7fb3f2199b50, 0xc820566a40, 0xc8200d0a80)
notaryserver_1 |        /go/src/github.com/docker/notary/utils/http.go:76 +0xe69
notaryserver_1 | net/http.(Handler).ServeHTTP-fm(0x7fb3f2199a58, 0xc820026130, 0xc8200d0a80)
notaryserver_1 |        /go/src/github.com/docker/notary/Godeps/_workspace/src/github.com/prometheus/client_golang/prometheus/http.go:61 +0x50
notaryserver_1 | github.com/prometheus/client_golang/prometheus.InstrumentHandlerFuncWithOpts.func1(0x7fb3f2199980, 0xc820537080, 0xc8200d0a80)
notaryserver_1 |        /go/src/github.com/docker/notary/Godeps/_workspace/src/github.com/prometheus/client_golang/prometheus/http.go:158 +0x335
notaryserver_1 | net/http.HandlerFunc.ServeHTTP(0xc820279450, 0x7fb3f2199980, 0xc820537080, 0xc8200d0a80)
notaryserver_1 |        /usr/local/go/src/net/http/server.go:1422 +0x3a
notaryserver_1 | github.com/gorilla/mux.(*Router).ServeHTTP(0xc8202787d0, 0x7fb3f2199980, 0xc820537080, 0xc8200d0a80)
notaryserver_1 |        /go/src/github.com/docker/notary/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
notaryserver_1 | net/http.serverHandler.ServeHTTP(0xc820267e60, 0x7fb3f2199980, 0xc820537080, 0xc8200d0a80)
notaryserver_1 |        /usr/local/go/src/net/http/server.go:1862 +0x19e
notaryserver_1 | net/http.(*conn).serve(0xc8205368f0)
notaryserver_1 |        /usr/local/go/src/net/http/server.go:1361 +0xbee
notaryserver_1 | created by net/http.(*Server).Serve
notaryserver_1 |        /usr/local/go/src/net/http/server.go:1910 +0x3f6
```

These errors only occur for repos in which I have previously pushed an image (where there is trust data already).  If I push to a different repository, everything works as expected until a restart of my notary compose services.  I'm not sure why this only occurs after a stop and re-start as the data in my database appears to still be intact.  Any thoughts on why this might be occurring or better methods of setting up notary with data persistence for longer term testing?

**Comment by endophage** [comment]
tl;dr; the fix is here: https://github.com/docker/notary/pull/323

Hi @mbentley, thanks for the report. I think the signer configuration must have changed during all the refactoring that went on prior to DockerCon. The clue is in the `(sql: Scan error on column index 1: unsupported driver -> Scan pair: []uint8  -> *time.Time)` errors (although one would have had to have seen it before to recognize the problem).

The ORM we're using, GORM, has 3 columns to track create/update/(soft)delete times and they are typed as `time.Time`. However, The Go MySQL Driver (at least the one we're using) returns the values as type `[]uint8` and the resulting error in type conversion results in the key not being "found".

The server, while also using GORM, doesn't cause any errors because all its queries pluck specific columns from the database tables, (unintentionally) avoiding any that would get assigned to a field of type `time.Time`.

The MySQL Driver we're using allows the setting of a query string parameter in the connection string to tell it to parse `DATE` and `DATETIME` MySQL columns in to `time.Time` and setting that parameter fixes the errors and allows the key to successfully be retrieved from the database.

We need to look at why we were allowing execution to proceed to a panic in the event the key isn't found, but the PR at the top (and accidentally submitted on your fork), will solve your immediate problem.

**Comment by mbentley** [comment]
Thanks! Much appreciated!

**Comment by endophage** [comment]
Issue about panic filed https://github.com/docker/notary/issues/324

**Comment by endophage** [comment]
@mbentley no worries. We never saw this issue because we have in memory caching of keys (hence no errors for a repo created and used within a single run of the containers). This was a good catch and I'm happy it's an easy fix :-)

**Comment by endophage** [comment]
@mbentley actually, I'd love it if you PR'd in your change to notarymysql/start. I don't want to enforce volumes in the compose file as we're intending that to create throwaway development instances, but for others that want to use persistence, having your changes in notarymysql/start would make it as intuitive as just adding the volume.

**Comment by mbentley** [comment]
https://github.com/docker/notary/pull/325, thanks!

**Comment by endophage** [comment]
Fix is now merged to master :-)
